hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/utils/PipelinesV2.java

```diff
@@ -0,0 +1,192 @@
+import org.apache.flink.streaming.api.datastream.DataStreamSink;
+import org.apache.flink.streaming.api.operators.ProcessOperator;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.planner.plan.nodes.exec.utils.ExecNodeUtil;
+import org.apache.flink.table.types.logical.RowType;
+
+import static org.apache.hudi.sink.utils.Pipelines.opUID;
+
+/**
+ * Utilities to generate pipelines for hudi sink V2.
+ */
+public class PipelinesV2 {
+
+  private static final String SINK_V2_NAME = "sink_v2";
+
+  public static DataStreamSink<RowData> sink(
+      DataStream<RowData> dataStream,
+      Configuration conf,
+      RowType rowType,
+      boolean overwrite,
+      boolean isBounded) {
+    HoodieSink hoodieSink = new HoodieSink(conf, rowType, overwrite, isBounded);
+    return dataStream.sinkTo(hoodieSink)
+        .setParallelism(getParallelismForSinkV2(conf))
+        .uid(opUID(SINK_V2_NAME, conf))
+        .name(SINK_V2_NAME);
+  }
+
+  public static DataStream<RowData> composePipeline(
+      DataStream<RowData> dataStream,
+      Configuration conf,
+      RowType rowType,
+      boolean overwrite,
+      boolean isBounded) {
+    // bulk_insert mode
+    if (OptionsResolver.isBulkInsertOperation(conf)) {
+      if (!isBounded) {
+        throw new HoodieException(
+            "The bulk insert should be run in batch execution mode.");
+      }
+      return Pipelines.bulkInsert(conf, rowType, dataStream);
+    }
+
+    // Append mode
+    if (OptionsResolver.isAppendMode(conf)) {
+      // close compaction for append mode
+      conf.set(FlinkOptions.COMPACTION_SCHEDULE_ENABLED, false);
+      DataStream<RowData> pipeline = Pipelines.append(conf, rowType, dataStream);
+      if (OptionsResolver.needsAsyncClustering(conf)) {
+        return clusterV2(conf, rowType, pipeline);
+      } else if (OptionsResolver.isLazyFailedWritesCleanPolicy(conf)) {
+        // add clean function to rollback failed writes for lazy failed writes cleaning policy
+        return cleanV2(conf, pipeline);
+      } else {
+        return pipeline;
+      }
+    }
+
+    // process dataStream and write corresponding files
+    DataStream<RowData> pipeline;
+    final DataStream<HoodieFlinkInternalRow> hoodieRecordDataStream = Pipelines.bootstrap(conf, rowType, dataStream, isBounded, overwrite);
+    pipeline = Pipelines.hoodieStreamWrite(conf, rowType, hoodieRecordDataStream);
+    // compaction
+    if (OptionsResolver.needsAsyncCompaction(conf)) {
+      // use synchronous compaction for bounded source.
+      if (isBounded) {
+        conf.set(FlinkOptions.COMPACTION_ASYNC_ENABLED, false);
+      }
+      return compactV2(conf, pipeline);
+    } else {
+      return cleanV2(conf, pipeline);
+    }
+  }
+
+  /**
+   * Get parallelism for Sink V2 writer to make sure the dummy writer operator can be chained with upstream operators.
+   */
+  private static int getParallelismForSinkV2(Configuration conf) {
```

Comment:
Is there an elegant way to infer the sink parallelism as the upstream operators?

Reply:
* `getParallelismForSinkV2` is getting parallelism for `HoodieSink#SinkWriter`
* the value is parallelism of the final operator of write pipeline which is build in `HoodieSink#addPreWriteTopology`. 

The write pipeline cannot be  pre-built outside `HoodieSink`, so it seems we cannot fetch the parallelism from operator/transformation directly.